Rewrite of filters + new filters

mundya opened the pull request:
- [x] Rewrite and clean up of filter code in C
- [x] General purpose LTI filter

mundya committed:
Remove unused filter code

mundya committed:
Rewrite of filter functionality

This commit makes some major changes to the ways filters are handled in
C.  Please refer to the code for detailed comments.  In brief:

  - A filter is a function which accepts a dimensionality, an input
    vector, an output vector and some filter-specific state.
  - Instances of filters are instantiated with function pointers which
    point at the function which should be used to evaluate them

The way the data is packed to be transmitted to the C has changed to
represent these changes. Each filter entry (as written out by the
Python) contains at least four words. The first word indicates the
number of extra words that are the parameters for the filter; the second
word indexes an array of function pointers to find the initialisation
method for the given filter; the third word indicates the "width" of the
filter (the number of dimensions in the input and output vector) and the
fourth word contains flags. Additional words contain filter-specific
parameters (e.g., coefficients, indication of order, etc.).

An additional test has been included to check that the filters work
correctly in C. **No new filter types have been implemented in this
commit.**

tcstewar commented:
> This is a prerequisite for removing passnodes, as this process can result in higher-order filters being created.

Just a quick comment on this -- technically, this shouldn't be a prerequisite for Spaun, in that all of the passthrough nodes in the system should only have a single filter on them.  Any situation where this isn't true is a bug in Spaun.

mundya commented:
@tcstewar I've only noticed combining filters being a problem when probing passthrough nodes in some models (e.g. SPA sequence).

mundya committed:
Add debug messages

tcstewar commented:
> when probing passthrough nodes in some models

A quick fix there is just to put synapse=None on that Probe.  Basically, if you're probing a passthrough node that's already taking a synapse into account, then you shouldn't filter it again.  So if any of the tests do that right now, they should be modified to not double-filter.

That said, in the long run it'd be nice to support this eventually.  :)

mundya commented:
> A quick fix there is just to put synapse=None on that Probe. Basically, if you're probing a passthrough node that's already taking a synapse into account, then you shouldn't filter it again. So if any of the tests do that right now, they should be modified to not double-filter.

Weirdly I found that even setting `synapse=None` on the utility probe in SPA sequence caused issues... not sure why that would be.

> That said, in the long run it'd be nice to support this eventually. :)

I think the rewrite as it stands is pleasant, but I don't think it'll be much more effort to add in 2nd order LPF (and general LTI support). I'm definitely hoping we can review this short to medium term :)

mundya committed:
Add LinearFilter support

 - Fix point rounding has a horrific effect on performance, this should
   be addressed in the future.

mundya commented:
@neworderofjamie - would you mind having a glance at the largely rewritten C?

neworderofjamie commented on nengo_spinnaker/regions/filters.py:
```diff
@@ -68,35 +82,31 @@ def __init__(self, filters, dt):
-        # 1 word + the size of the largest supported filter * number of filters
-        return 4 + self._largest_struct * len(self.filters)
+        # 1 word + the size of the each filter
+        words = (1 + 4*len(self.filters) +
```
what're the `4 * len(self.filters)` words for?

neworderofjamie commented on spinnaker_components/common/input_filtering.c:
```diff
@@ -0,0 +1,320 @@
+  use(params);
+
+  // The None filter just copies its input to the output
+  for (uint32_t d = 0; d < n_dims; d++)
```
IMHO, there is never a good reason not to use curly brackets - more vertical space is always a good thing

mundya replied:
Done.

neworderofjamie commented on spinnaker_components/common/input_filtering.c:
```diff
@@ -0,0 +1,320 @@
+}
+
+/* 1st Order Low-Pass ********************************************************/
+struct _lowpass_state
```
If you typedef it then you don't need all the struct horror lower down

mundya replied:
I personally don't find that horrible, but I'll switch it.

neworderofjamie commented on spinnaker_components/common/input_filtering.c:
```diff
@@ -0,0 +1,320 @@
+  for (uint32_t d = 0; d < n_dims; d++)
+  {
+    output[d] *= params->a;
+    output[d] += input[d] * params->b;
```
The ideal solution generated code wise would be for this to generate:

```
SMULL out_lo, out_hi, a, out 
SMLAL out_lo, out_hi, in, b
```

Which would result in an answer in a result in S33.30 spread across out_lo and out_hi. Then you could munge together the 32 bits from the middle of out_hi and out_lo and obtain the correct S16.15 result )

mundya replied:
I suggest we fiddle with inlining some assembly when we look at changing the fixed point.  Noted that there are lots of places where a `SMLAL` would be a good thing!

neworderofjamie commented on spinnaker_components/common/input_filtering.c:
```diff
@@ -0,0 +1,320 @@
+
+  // Copy the filter parameters into memory
+  MALLOC_OR_DIE(filter->state, sizeof(struct _lowpass_state));
+  spin1_memcpy(filter->state, params, sizeof(struct _lowpass_state));
```
To minimise code size, it might be worth checking we standardise on whichever is smaller of memcpy and spin1_memcpy.

neworderofjamie commented on spinnaker_components/common/input_filtering.c:
```diff
@@ -0,0 +1,320 @@
+    for (int32_t k=0; k < (int32_t) state->order; k++)
+    {
+      output[d] -= state->a[k] * y[(state->n - k - 1) % state->order];
+      output[d] += state->b[k] * x[(state->n - k - 1) % state->order];
```
% by runtime value is going to result in a software divide. I would say the options are, store power of two number of, (potentially zeroed coefficients) and mask them or seek some sort of reason within https://github.com/SpiNNakerManchester/spinn_common/blob/master/src/circular_buffer.

mundya replied:
I can do subtraction and an if, I was just being lazy... I keep forgetting to avoid divide.

neworderofjamie commented on spinnaker_components/common/input_filtering.c:
```diff
@@ -0,0 +1,320 @@
+
+    // Save the current output for later steps
+    y[state->n] = output[d];
+  }
```
Also, if you put a, b, y and x in a struct, they'd be known distances in memory apart unlike if they're in separate arrays specified by pointers so a lot of this addressing stuff would probably optimise out.

neworderofjamie commented on spinnaker_components/common/input_filtering.c:
```diff
@@ -0,0 +1,320 @@
+  }
+
+  // Rotate the ring buffer.
+  state->n = (state->n + 1) % state->order;
```
Again, here be divides

neworderofjamie commented on spinnaker_components/common/input_filtering.c:
```diff
@@ -0,0 +1,320 @@
+
+  // If debugging then print out all filter parameters
+#ifdef DEBUG
+  for (uint32_t k = 0; k < state->order; k++)
```
{

mundya replied:
Done

neworderofjamie commented on spinnaker_components/common/input_filtering.c:
```diff
@@ -0,0 +1,320 @@
+#ifdef DEBUG
+  for (uint32_t k = 0; k < state->order; k++)
+    io_printf(IO_BUF, "  a[%d] = %k\n", k, state->a[k]);
+
```
}

neworderofjamie commented on spinnaker_components/common/input_filtering.c:
```diff
@@ -0,0 +1,320 @@
+    io_printf(IO_BUF, "  a[%d] = %k\n", k, state->a[k]);
+
+  for (uint32_t k = 0; k < state->order; k++)
+    io_printf(IO_BUF, "  b[%d] = %k\n", k, state->b[k]);
```
}

neworderofjamie commented on spinnaker_components/common/input_filtering.c:
```diff
@@ -0,0 +1,320 @@
+  for (uint32_t k = 0; k < state->order; k++)
+    io_printf(IO_BUF, "  a[%d] = %k\n", k, state->a[k]);
+
+  for (uint32_t k = 0; k < state->order; k++)
```
{

neworderofjamie commented on spinnaker_components/common/input_filtering.c:
```diff
@@ -0,0 +1,320 @@
+  for (uint32_t n = 0; n < size * state->order; n++)
+  {
+    state->xz[n] = state->yz[n] = 0.0k;
+  }
```
memset? I'm sure it's called somewhere already and hence linked in - speed matters not at this point

neworderofjamie commented on spinnaker_components/common/input_filtering.c:
```diff
@@ -0,0 +1,320 @@
+{
+  uint32_t order;
+  value_t data;  // Array of parameters 2*order longer (a[...] || b[...])
+};
```
typedef to avoid more struct crud below

mundya replied:
Done

neworderofjamie commented on spinnaker_components/common/input_filtering.h:
```diff
@@ -0,0 +1,225 @@
+  void *state;              // State maintained by the filter
+
+  FilterStep step;  // Filter evaluation function
+};
```
Typedef again

neworderofjamie commented on spinnaker_components/common/input_filtering.h:
```diff
@@ -0,0 +1,225 @@
+    struct input_filtering_collection *filters)
+{
+  // Apply the filter step for each filter in the collection.
+  for (uint32_t n = 0; n < filters->n_filters; n++)
```
{

neworderofjamie commented on spinnaker_components/common/input_filtering.h:
```diff
@@ -0,0 +1,225 @@
+  // Apply the filter step for each filter in the collection.
+  for (uint32_t n = 0; n < filters->n_filters; n++)
+    _if_filter_step(&filters->filters[n]);
+}
```
}

neworderofjamie commented on spinnaker_components/common/input_filtering.c:
```diff
@@ -0,0 +1,320 @@
+    {
+      filters->filters[f].input->value[d] = 0.0k;
+      filters->filters[f].output[d] = 0.0k;
+    }
```
memset again maybe

mundya commented:
@neworderofjamie would you mind taking another quick glance? I'd like to squash before merging anyways...

mundya committed:
Merge branch 'master' into new-filters

Conflicts:
	nengo_spinnaker/binaries/nengo_ensemble.aplx
	nengo_spinnaker/binaries/nengo_filter.aplx
	nengo_spinnaker/binaries/nengo_mc_player.aplx
	nengo_spinnaker/binaries/nengo_rx.aplx
	nengo_spinnaker/binaries/nengo_tx.aplx
	nengo_spinnaker/binaries/nengo_value_sink.aplx
	nengo_spinnaker/binaries/nengo_value_source.aplx
	spinnaker_components/ensemble/Makefile
	spinnaker_components/ensemble/ensemble_harness.c
	spinnaker_components/ensemble/ensemble_update.c
	spinnaker_components/filter/Makefile
	spinnaker_components/sdp_tx/Makefile
	spinnaker_components/value_sink/Makefile

mundya commented:
@neworderofjamie - were we sufficiently happy with this to merge?

mundya committed:
Merge branch 'master' into new-filters
